src/compiler/graph/link_nodes.rs

```diff
@@ -0,0 +1,270 @@
+// TODO: remove this
+#![allow(unused, dead_code)]
+
+use paste::paste;
+
+use crate::types::{
+    block::{redstone::Component, Block, Facing, Kind},
+    compiler::{Graph, Link},
+    contraption::{Position, World},
+};
+
+pub fn link_nodes<'a, W: World<'a>>(graph: Graph, world: &'a W) -> Graph {
+    let indices = graph.node_indices();
+
+    for index in indices {
+        let node = graph.node_weight(index).expect("node should exist");
+        let block = world.get_block(node.pos).expect("block should exist");
+
+        let sources = get_potential_sources(block, world);
+    }
+    todo!()
+}
+
+fn get_potential_sources<'a, W: World<'a>>(
+    block: &'a Block,
+    world: &'a W,
+) -> Vec<(Position, Link)> {
+    match block.get_kind() {
+        Kind::Block => get_sources(block, world, block::get_adjacent_source),
+        Kind::Component(Component::Lamp) => get_sources(block, world, lamp::get_adjacent_source),
+        _ => vec![],
+    }
+}
+
+pub fn get_sources<'a, W: World<'a>>(
+    block: &'a Block,
+    world: &'a W,
+    check_block_source: impl Fn(&'a Block, &'a Block) -> Option<(Position, Link)>,
+) -> Vec<(Position, Link)> {
+    let position = block.get_vec_pos();
+
+    let mut sources = vec![];
+
+    let mut add_state = (0, 1);
+    while add_state.0 < 3 {
+        if add_state.1 < -1 {
+            add_state = (add_state.0 + 1, 1);
+        }
+
+        let mut position = position.clone();
+        position[add_state.0] += add_state.1;
+
+        if let Some(adjacent_block) = world.get_block((position[0], position[1], position[2])) {
+            if let Some(source) = check_block_source(block, adjacent_block) {
+                sources.push(source);
+            }
+        };
+
+        add_state.1 -= 2;
+    }
+    sources
+}
+
+mod block {
+    use super::*;
+
+    pub fn get_adjacent_source(
+        current_block: &Block,
+        adjacent_block: &Block,
+    ) -> Option<(Position, Link)> {
+        let required_facing =
+            utils::get_facing(current_block.get_vec_pos(), adjacent_block.get_vec_pos())
+                .expect("should not be the same block");
+        match adjacent_block.get_kind() {
+            Kind::Component(Component::Dust) => {
+                if adjacent_block.get_facing().contains(&required_facing)
+                    || required_facing == Facing::NegativeY

```

Comment:
could we not put these if checks in the pat matching? ie `Kind::Component(...) if ... => {}.
This would allow us to match every case and not have to use return. and just do `_ => None`